File: app/src/main/java/com/pavelprymak/propodcast/presentation/screens/filters/RegionFilterFragment.java
package com.pavelprymak.propodcast.presentation.screens.filters;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.pavelprymak.propodcast.App;
import com.pavelprymak.propodcast.R;
import com.pavelprymak.propodcast.databinding.FragmentRegionFilterBinding;
import com.pavelprymak.propodcast.presentation.adapters.RegionAdapter;
import com.pavelprymak.propodcast.presentation.adapters.RegionClickListener;
import com.pavelprymak.propodcast.presentation.viewModels.RegionViewModel;

import static com.pavelprymak.propodcast.utils.SettingsPreferenceManager.ALL_REGIONS;


public class RegionFilterFragment extends Fragment implements RegionClickListener {
    private FragmentRegionFilterBinding mBinding;
    private RegionAdapter mAdapter;
    private RegionViewModel mRegionViewModel;

    public RegionFilterFragment() {
    }

    static RegionFilterFragment newInstance() {
        Bundle arg = new Bundle();
        RegionFilterFragment fragment = new RegionFilterFragment();
        fragment.setArguments(arg);
        return fragment;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        mRegionViewModel = ViewModelProviders.of(this).get(RegionViewModel.class);
        mBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_region_filter, container, false);
        return mBinding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        prepareRecycler();
        mRegionViewModel.getRegions().observe(this, regionItems -> {
            if (regionItems != null) {
                String selectedRegionShortName = App.mSettings.getFilterRegion();
                mAdapter.updateList(regionItems, selectedRegionShortName);
                if (!selectedRegionShortName.equals(ALL_REGIONS)) {
                    int selectedItemPosition = mAdapter.getPositionByRegionShortName(selectedRegionShortName);
                    mBinding.recyclerRegions.smoothScrollToPosition(selectedItemPosition);
                }
            }
        });
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mRegionViewModel.getRegions().removeObservers(this);
    }

    private void prepareRecycler() {
        mAdapter = new RegionAdapter(this);
        LinearLayoutManager layoutManager = new LinearLayoutManager(getContext(), RecyclerView.VERTICAL, false);
        mBinding.recyclerRegions.setLayoutManager(layoutManager);
        mBinding.recyclerRegions.setHasFixedSize(true);
        mBinding.recyclerRegions.setAdapter(mAdapter);
    }

    @Override
    public void onRegionItemClick(String regionName) {
        App.mSettings.saveFilterRegion(regionName);
    }
}

File: app/src/main/java/com/pavelprymak/propodcast/presentation/screens/filters/RegionFilterFragment.kt
package com.pavelprymak.propodcast.presentation.screens.filters

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.pavelprymak.propodcast.App
import com.pavelprymak.propodcast.R
import com.pavelprymak.propodcast.presentation.adapters.RegionAdapter
import com.pavelprymak.propodcast.presentation.adapters.RegionClickListener
import com.pavelprymak.propodcast.presentation.viewModels.RegionViewModel
import com.pavelprymak.propodcast.utils.SettingsPreferenceManager
import com.pavelprymak.propodcast.utils.SettingsPreferenceManager.ALL_REGIONS
import com.pavelprymak.propodcast.utils.otto.filters.EventUpdateRegionFilter
import kotlinx.android.synthetic.main.fragment_region_filter.*
import org.koin.android.ext.android.inject
import org.koin.android.viewmodel.ext.android.viewModel


class RegionFilterFragment : Fragment(), RegionClickListener {
    private var mAdapter: RegionAdapter? = null
    private val mRegionViewModel: RegionViewModel by viewModel()
    private val mSettingsPref: SettingsPreferenceManager by inject()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_region_filter, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        prepareRecycler()
        mRegionViewModel.regions.observe(this, Observer { regionItems ->
            if (regionItems != null) {
                val selectedRegionShortName = mSettingsPref.filterRegion
                mAdapter?.updateList(regionItems, selectedRegionShortName)
                if (selectedRegionShortName != ALL_REGIONS) {
                    val selectedItemPosition = mAdapter?.getPositionByRegionShortName(selectedRegionShortName)
                    recyclerRegions.smoothScrollToPosition(selectedItemPosition ?: 0)
                }
            }
        })
    }

    override fun onDestroyView() {
        super.onDestroyView()
        mRegionViewModel.regions.removeObservers(this)
    }

    private fun prepareRecycler() {
        mAdapter = RegionAdapter(this)
        val layoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)
        recyclerRegions.layoutManager = layoutManager
        recyclerRegions.setHasFixedSize(true)
        recyclerRegions.adapter = mAdapter
    }

    override fun onRegionItemClick(regionName: String) {
        mSettingsPref.saveFilterRegion(regionName)
        if (resources.getBoolean(R.bool.isTablet)) {
            App.eventBus.post(EventUpdateRegionFilter())
        }
    }

    companion object {
        internal fun newInstance(): RegionFilterFragment {
            val arg = Bundle()
            val fragment = RegionFilterFragment()
            fragment.arguments = arg
            return fragment
        }
    }
}
